cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,805 @@
+                                                     unique_pair_counts.begin())),
+        thrust::make_zip_iterator(thrust::make_tuple(std::get<0>(unique_pairs_buffer).end(),
+                                                     std::get<1>(unique_pairs_buffer).end(),
+                                                     unique_pair_counts.end())),
+        [] __device__(auto val) { return thrust::get<1>(val); },
+        handle.get_stream());
+
+    //
+    // Count local #elments for all the types mapped to this GPU
+    //
+
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        std::get<0>(unique_pairs_buffer).begin(),
+                        std::get<0>(unique_pairs_buffer).end(),
+                        unique_pair_counts.begin());
+
+    auto nr_unique_types = thrust::count_if(
+      handle.get_thrust_policy(),
+      thrust::make_counting_iterator(size_t{0}),
+      thrust::make_counting_iterator(std::get<0>(unique_pairs_buffer).size()),
+      detail::is_first_in_run_t<edge_type_t const*>{std::get<0>(unique_pairs_buffer).data()});
+
+    unique_types.resize(static_cast<size_t>(nr_unique_types), handle.get_stream());
+    unique_type_counts.resize(static_cast<size_t>(nr_unique_types), handle.get_stream());
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          std::get<0>(unique_pairs_buffer).begin(),
+                          std::get<0>(unique_pairs_buffer).end(),
+                          unique_pair_counts.begin(),
+                          unique_types.begin(),
+                          unique_type_counts.begin());
+
+  } else {
+    auto edge_types = cugraph::extract_transform_e(
+      handle,
+      graph_view,
+      cugraph::edge_src_dummy_property_t{}.view(),
+      cugraph::edge_dst_dummy_property_t{}.view(),
+      edge_type_view,
+      cuda::proclaim_return_type<thrust::optional<edge_type_t>>(
+        [] __device__(auto, auto, thrust::nullopt_t, thrust::nullopt_t, edge_type_t et) {
+          return thrust::optional<edge_type_t>{et};
+        }));
+
+    thrust::sort(handle.get_thrust_policy(), edge_types.begin(), edge_types.end());
+
+    auto nr_unique_types =
+      thrust::count_if(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator(size_t{0}),
+                       thrust::make_counting_iterator(edge_types.size()),
+                       detail::is_first_in_run_t<edge_type_t const*>{edge_types.data()});
+
+    unique_types.resize(static_cast<size_t>(nr_unique_types), handle.get_stream());
+    unique_type_counts.resize(static_cast<size_t>(nr_unique_types), handle.get_stream());
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          edge_types.begin(),
+                          edge_types.end(),
+                          thrust::make_constant_iterator(size_t{1}),
+                          unique_types.begin(),
+                          unique_type_counts.begin());
```

Comment:
Same here. copy_if + binary_search.

Reply:
Oh, and thrust has unique_copy as well. This will be a better choice.

https://nvidia.github.io/cccl/thrust/api/groups/group__stream__compaction.html#function-unique-copy

Reply:
changed to copy_if + binary_search.